src/Features/Core/Portable/Extensions/ExtensionMessageHandlerService.cs

```diff
@@ -162,28 +183,44 @@ private async ValueTask<ExtensionMessageNames> GetExtensionMessageNamesInCurrent
 
         private async ValueTask<ExtensionMessageResult> HandleExtensionMessageInCurrentProcessAsync<TArgument>(
             TArgument executeArgument, bool isSolution, string messageName, string jsonMessage,
-            Dictionary<string, AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>>> cachedHandlers,
+            CachedHandlers cachedHandlers,
             CancellationToken cancellationToken)
         {
-            AsyncLazy<ImmutableArray<IExtensionMessageHandlerWrapper>> lazyHandlers;
+            AsyncLazy<HandlerWrappers> lazyHandlers;
+            var potentiallyRefersToUnregisteredHandlerName = false;
             lock (_gate)
             {
                 // May be called a lot.  So we use the non-allocating form of this lookup pattern.
                 lazyHandlers = cachedHandlers.GetOrAdd(
                     messageName,
                     static (messageName, arg) => AsyncLazy.Create(
-                        static (arg, cancellationToken) => arg.@this.ComputeHandlersAsync(arg.messageName, arg.isSolution, cancellationToken),
-                        (messageName, arg.@this, arg.isSolution)),
+                        static (arg, cancellationToken) => arg.@this.ComputeHandlers(arg.messageName, arg.isSolution, cancellationToken),
+                        arg),
                     (messageName, @this: this, isSolution));
+
+                var unregisteredHandlerNames = isSolution
+                    ? _unregisteredHandlerNames_useOnlyUnderLock.workspace
+                    : _unregisteredHandlerNames_useOnlyUnderLock.document;
+                potentiallyRefersToUnregisteredHandlerName = unregisteredHandlerNames.Contains(messageName);
             }
 
             var handlers = await lazyHandlers.GetValueAsync(cancellationToken).ConfigureAwait(false);
 
-            // Throwing here indicates a bug in the gladstone client itself (as it allowed calling into an lsp message
-            // that has no registered handlers).  So we want this to bubble outwards as a failure that disables
-            // extension running in the OOP process.  This must be fixed by gladstone.
             if (handlers.Length == 0)
+            {
+                // It's ok to find no handlers *if* this was the name of a handler that was previously unloaded. As
+                // gladstone allows unloads to happen concurrently with calls to handlers, it's possible that we may
+                // unload first, then receive a call to the handler.  In that case, just report back to the client what
+                // happened and let them decide what to do.
+                if (potentiallyRefersToUnregisteredHandlerName)
```

Comment:
This all seems relatively complicated just to support us being called concurrently for a handler that got unloaded.

I'm wondering if it would be reasonable to instead not throw if we get a request for a non-existent handler - but return an exception as data.  While technically not a plugin failure, it also doesn't necessarily mean the service failed.  But rather the client sent us an invalid request.  That would avoid having to do the cahcing of unloading handlers right?

Not sure if its better though

Reply:
i want to avoid really broken things.  like resetting, then calling in for a handler that doesn't exist.  :)

i do hear you on complexity.  but i think we want to really be strict up front.